misc/dbt-materialize/dbt/adapters/materialize/impl.py

```diff
@@ -178,3 +179,75 @@ def render_raw_columns_constraints(
                 rendered_column_constraints.append(" ".join(rendered_column_constraint))
 
         return rendered_column_constraints
+
+    def get_column_schema_from_query(self, sql: str) -> List[PostgresColumn]:
+        # The idea is that this function returns the names and types of the
+        # columns returned by `sql` without actually executing the statement.
+        #
+        # The easiest way to do this would be to create a prepared statement
+        # using `sql` and then inspect the result type of the prepared
+        # statement. Unfortunately the Python DB-API and the underlying psycopg2
+        # driver do not support preparing statements.
+        #
+        # So instead we create a temporary view based on the SQL statement and
+        # inspect the types of the columns in the view using `mz_columns`.
+        #
+        # Note that the upstream PostgreSQL adapter takes a different approach.
+        # It executes `SELECT * FROM (<query>) WHERE FALSE LIMIT 0`, which in
+        # PostgreSQL is optimized to a no-op that returns no rows, and then
+        # inspects the description of the returned cursor. That doesn't work
+        # well in Materialize, because `SELECT ... LIMIT 0` requires a cluster
+        # (even though it returns no rows), and we're not guaranteed that the
+        # connection has a valid cluster.
+
+        # Determine the name to use for the view. Unfortunately Materialize
+        # comingles data about temporary views from all sessions in the system
+        # catalog, so we need to manually include the connection ID in the name
+        # to be able to identify the catalog entries for the temporary views
+        # created by this specific connection.
+        connection_id = self.connections.get_thread_connection().handle.info.backend_pid
+        view_name = f"__dbt_sbq{connection_id}"
+
+        # NOTE(morsapaes): because we need to consider the existence of a
+        # sql_header configuration (see dbt-core #7714) but don't support
+        # running the header SQL in the same transaction as the model SQL
+        # statement, we split the input based on the string appended to the
+        # header in materialize__get_empty_subquery_sql.
```

Comment:
The one downside of this approach is it _can_ be broken if someone somehow includes the `"#__dbt_sbq_parse_header__#"` string in their SQL header. But if you feel better admitting this tiny edge case for simpler code, great, let's go with it.

Reply:
I went back and forth on it, and considered using something safer like `{{ invocation_id }}`, but...dbt is so tough to debug that I'd rather lean towards more straightforward code than optimize for the 0.0000009% chance someone does include the string in their SQL.